lib/chart/animated_line_chart.dart

```diff
@@ -631,6 +700,90 @@ class ChartPainter extends CustomPainter {
     }
   }
 
+  void _drawLegends(Size size, Canvas canvas) {
+    assert(legends!.length ==
+            _chart.lines.where((line) => line.isMarkerLine).length ||
+        legends!.length ==
+            _chart.lines.where((line) => line.isMarkerLine == false).length ||
+        legends!.length == _chart.lines.length);
+
+    List<double> values = [];
+
+    if (_chart.seriesMap != null) {
+      _chart.seriesMap!.forEach((key, value) {
+        if (key == 0) {
+        } else {
+          value.forEach((highlightPoint) {
+            values.add(highlightPoint.chartPoint.y);
+          });
+        }
+      });
+    }
+
+    List<double> distinctValues = values.toSet().toList();
+
+    for (int i = legends!.length - 1; i >= 0; i--) {
+      final textStyle = legends?[i].style ??
+          TextStyle(
+            color: Colors.black,
+            fontSize: 10,
+            overflow: TextOverflow.clip,
+          );
+      final textSpan = TextSpan(
+        text: legends?[i].title,
+        style: textStyle.copyWith(fontSize: 10),
+      );
+      final textPainter = TextPainter(
+        text: textSpan,
+        textAlign: TextAlign.left,
+        textDirection: TextDirectionHelper.getDirection(),
+      );
+      textPainter.layout(
+        minWidth: 0,
+        maxWidth: size.width,
+      );
+
+      if (i > 0 &&
+          distinctValues[i] - distinctValues[i - 1] <= 10 &&
+          distinctValues[i] - distinctValues[i - 1] >= 0) {
+        distinctValues[i] = distinctValues[i] + 3;
+        distinctValues[i - 1] = distinctValues[i - 1] - 3;
+      }
+
+      if (legends![i].showLeadingLine!) {
+        canvas.drawLine(
+            Offset(size.width + 5, distinctValues[i]),
+            Offset(size.width + 15, distinctValues[i]),
+            Paint()
+              ..color = legends![i].color ?? Colors.blue
+              ..style = PaintingStyle.stroke
+              ..strokeWidth = 2);
+      }
+
+      if (legends![i].icon != null) {
+        final iconPainter = TextPainter(
+          textAlign: TextAlign.left,
+          textDirection: TextDirectionHelper.getDirection(),
+        );
+
+        iconPainter.text = TextSpan(
+          text: String.fromCharCode(legends![i].icon!.icon!.codePoint),
+          style: TextStyle(
+            fontSize: 12.0,
+            fontFamily: legends![i].icon!.icon!.fontFamily,
```

Comment:
Is `icon` non null here ?

Reply:
`icon!.icon!` ?

Reply:
Yes this icon!.icon! is quite silly. I am taking the Icon from the Legend and Icon has something called IconData, which can be reached by saying Icon.icon, which in this case becomes icon!.icon! because both Icon is nullable and IconData from the Icon class is nullable.